### PR TITLE
[MXNET-1001] Add new cuddn flag in Gluon Conv blocks to control - cudnn_tune, cudnn_off params

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
## Description ##
<s>Add support for two flags - `cudnn_tune` and `cudnn_off` for conv blocks in Gluon.
These options are already supported in symbolic APIs. In this change, we basically take that as param and pass it to underlying ndarray/symbolic APIs. </s>

Add new `cudnn` flag in Gluon Conv blocks to control underlying `cudnn_tune` and `cudnn_off` parameters. Note that, these are supported in sym and ndarray APIs for Conv.

fixes https://github.com/apache/incubator-mxnet/issues/11342

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Support cudnn flag in gluon conv blocks that under the hood controls cudnn_tune and cudnn_off params for sym/ndarray conv APIs.
- [X] Add tests to test cudnn_tune and cudnn_off in gluon conv blocks

### Comments ###
- Verify if API docs are generated correctly as expected.

### CODE REVIEW

**Comment by sbodenstein** [comment]
@szha: I do agree in general. But with cuDNN, you sometimes want to disable it only for specific layers. For example, we disable cuDNN for batchnorm layers when `eps < 10^-5`, as cuDNN fails in this case (and sometimes you import nets from other frameworks with smaller values for `eps` that you want to run). You definitely don't want a global variable switching off cuDNN for all layers in the net.

File: python/mxnet/gluon/nn/conv_layers.py [hunk 1]
```diff
@@ -86,10 +86,15 @@ class _Conv(HybridBlock):
         Initializer for the `weight` weights matrix.
     bias_initializer: str or `Initializer`
         Initializer for the bias vector.
+    cudnn_tune: {None, 'fastest', 'limited_workspace', 'off'}, Optional, default='None'
+        Whether to pick convolution algo by running performance test.
+    cudnn_off: boolean, optional, default=False
```
[the remark this answers is not in the document]
Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/nn/conv_layers.py]: Thanks Naveen for the review. Yes, I will add details for all the options.
cudnn_off = False => Will ignore cudnn_tune and does not use cudnn for this layer. I will update the document to reflect this.

cudnn_tune=None will use environment variable MXNET_CUDNN_AUTOTUNE_DEFAULT value. By default this environment variable value will be '1' i.e., limited_workspace i.e., choose algo based on memory availability.

**Comment by sandeep-krishnamurthy** [comment]
@szha - I agree, but, we need to expose the control for users as they have preference on these choices of underlying algorithm.

I am thinking of standardizing the API with one parameter for cudnn controls. 
```
conv( channels, ..., cudnn={‘off’, 'default', ‘fastest’, ‘limited_workspace’, ‘global’})

default: limited_workspace

off: Do not use cudnn for this layer.
default: Use Cudnn. Do not tune.
limited_workspace: Use Cudnn. Run the test and pick the fastest algorithm that doesn’t exceed workspace limit.
fastest: Use Cudnn. Pick the fastest algorithm and ignore workspace limit.
global: Use Cuddn. Use global Cudnn behavior determined by the environment variable MXNET_CUDNN_AUTOTUNE_DEFAULT. 0 for off, 1 for limited workspace (default), 2 for fastest.
```

@nswamy @szha @sbodenstein - Let me know what you guys think. I will make corresponding changes.

**Comment by vrakesh** [comment]
@sandeep-krishnamurthy Requesting an update on the status of this PR, please look into the build failures and feedback

**Comment by sandeep-krishnamurthy** [comment]
@szha @nswamy @zhreshold - Requesting for review.

File: python/mxnet/gluon/nn/conv_layers.py [hunk 2]
```diff
@@ -86,10 +121,20 @@ class _Conv(HybridBlock):
         Initializer for the `weight` weights matrix.
     bias_initializer: str or `Initializer`
         Initializer for the bias vector.
+    cudnn : str,
+        Supported values - {'off', 'default', 'fastest', 'limited_workspace', 'global'}.
```
Comment (zhreshold) [inline python/mxnet/gluon/nn/conv_layers.py]: by default it should align with env variable, i.e., the global in your revision. And I don't quite like the name global, sounds like a global tuning option. Is env a better name?

**Comment by ankkhedia** [comment]
@sandeep-krishnamurthy ping!
 Could you please address the review comments and fix the branch conflicts.

File: python/mxnet/gluon/nn/conv_layers.py (hunk 2 above)
Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/nn/conv_layers.py]: Thanks.
default -> is actual default now i.e., uses environment variable.
no_autotune -> Uses default cudnn algorithm.

So basically,
global is renamed to default,
default is renamed to no_autotune.

@zhreshold - Let me know if changes looks good. Thanks.

**Comment by sandeep-krishnamurthy** [comment]
Sorry for delayed response. Updated the PR.

**Comment by sandeep-krishnamurthy** [comment]
@zhreshold addressed your comments.

File: tests/python/gpu/test_gluon_gpu.py [hunk 3]
```diff
@@ -408,6 +431,60 @@ def tensor_size(big_tensor_bytes):
         # Evaluate model
         net(data_in).asnumpy()
 
+@with_seed()
+def test_conv_invalid_cudnn_flag():
+    try:
+        _check_layer_forward(nn.Conv1D(16, 3, in_channels=4, cudnn='fail_me'), (1, 4, 10))
+        assert False, "Gluon Conv1D block did not raise the AssertionError for invalid cudnn flag."
+    except AssertionError:
```
Comment (zhreshold) [inline tests/python/gpu/test_gluon_gpu.py]: Use  `AssertRaises`

Comment (sandeep-krishnamurthy) [reply tests/python/gpu/test_gluon_gpu.py]: Thanks. updated.

**Comment by sandeep-krishnamurthy** [comment]
Closing this for now, while we come up with a better approach.
